File: frontend/src/features/comparisons/types/comparison.ts
import type {
  PromptAnalysis,
} from "../../analysis/types/analysis";

export interface PromptComparisonRequest {
  left_prompt_id: number;
  right_prompt_id: number;

  left_variables: Record<string, string>;
  right_variables: Record<string, string>;
}

export interface PromptComparisonSummary {
  character_difference: number;
  word_difference: number;
  line_difference: number;
  token_difference: number;
  variable_difference: number;

  similarity: number;
}

export interface PromptComparisonResponse {
  left: PromptAnalysis;
  right: PromptAnalysis;

  summary: PromptComparisonSummary;
}

File: backend/app/schemas/comparison.py
from pydantic import BaseModel, Field

from app.schemas.analysis import PromptAnalysis


class PromptComparisonRequest(BaseModel):
    left_prompt_id: int = Field(gt=0)
    right_prompt_id: int = Field(gt=0)

    left_variables: dict[str, str] = Field(default_factory=dict)
    right_variables: dict[str, str] = Field(default_factory=dict)


class PromptComparisonSummary(BaseModel):
    character_difference: int
    word_difference: int
    line_difference: int
    token_difference: int
    variable_difference: int


class PromptComparisonResponse(BaseModel):
    left: PromptAnalysis
    right: PromptAnalysis
    summary: PromptComparisonSummary